tools/rv_histogram.c

```diff
@@ -241,39 +239,37 @@ int main(int argc, char *args[])
     }
 
     struct Elf32_Ehdr *hdr = get_elf_header(e);
-    struct Elf32_Shdr **shdrs = get_elf_section_headers(e);
-    if (!shdrs) {
-        fprintf(stderr, "malloc for section headers failed\n");
-        return 1;
-    }
+    uint8_t *elf_first_byte = get_elf_first_byte(e);
+    const struct Elf32_Shdr **shdrs =
+        (const struct Elf32_Shdr **) &elf_first_byte[hdr->e_shoff];
 
-    uintptr_t *elf_first_byte = (uintptr_t *) get_elf_first_byte(e);
     rv_insn_t ir;
     bool res;
 
     for (int i = 0; i < hdr->e_shnum; i++) {
-        struct Elf32_Shdr *shdr = (struct Elf32_Shdr *) shdrs + i;
+        const struct Elf32_Shdr *shdr = (const struct Elf32_Shdr *) &shdrs[i];
+        bool is_prg = shdr->sh_type & SHT_PROGBITS;
+        bool has_insn = shdr->sh_flags & SHF_EXECINSTR;
 
-        if (!(shdr->sh_type & SHT_PROGBITS && shdr->sh_flags & SHF_EXECINSTR))
+        if (!(is_prg && has_insn))
             continue;
 
-        uintptr_t *exec_start_addr =
-            (uintptr_t *) ((uint8_t *) elf_first_byte + shdr->sh_offset);
-        uintptr_t *exec_end_addr =
-            (uintptr_t *) ((uint8_t *) exec_start_addr + shdr->sh_size);
-        uintptr_t *ptr = (uintptr_t *) exec_start_addr;
+        uint8_t *exec_start_addr = &elf_first_byte[shdr->sh_offset];
+        const uint8_t *exec_end_addr = &exec_start_addr[shdr->sh_size];
+        uint8_t *ptr = exec_start_addr;
         uint32_t insn;
 
         while (ptr < exec_end_addr) {
 #if RV32_HAS(EXT_C)
             if ((*((uint32_t *) ptr) & FC_OPCODE) != 0x3) {
                 insn = *((uint16_t *) ptr);
-                ptr = (uintptr_t *) ((uint8_t *) ptr + 2);
+                ptr += 2;
```

Comment:
Can you confirm if `ptr += 2` is equivalent to the original form?

Reply:
ditto